### PR TITLE
Fix: Fixed an issue where the conflicts dialog didn't append numbers when pasting files

### LINKED ISSUE #11697: Bug: Generating new name for conflict is appending "copy" instead of a number "(#)"
### What feature or improvement do you think would benefit Files?

When I move or copy a file from one folder to another folder that already contains a file with the same name, and I would like to keep both, I would like the option for it to append " (#)" to the file (like Windows) instead of "- Copy" when I select Generate New Name. 

If it were the same file, I would almost always choose Replace or Skip, not Keep Both (aka Generate New Name), so when I choose to keep both, it is almost always because it isn't the same file, and therefore is not a copy, but an additional item with the same name.

### Requirements

- [ ] When copying and pasting items within the same folder, "- copy" should be appended (as is File Explorer)
- [ ] When copying and pasting item from a different folder, selecting generate new name from the conflict modal should append a number (as in File Explorer)

### Files Version

2.4.40.0

### Windows Version

Windows 11 22H2 22623.1020

### Comments

_No response_

### PR DESCRIPTION
**Pull request by ashrafmansuri**
**Resolved / Related Issues**
- [ ] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
   Close  #11697

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?
- [ ] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [ ] Did you implement any design changes to an existing feature?
   - [ ] Was this change approved?
- [x] Are there any other steps that were used to validate these changes?
   1. Open app ...
   2. Copy a file or folder from a similar source and destination the filename has -Copy and file behaves similar to Explorer ...
   3. Now copy a file or folder from a source different than destination the filename has (#) appended to it.

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

**Comment by ashrafmansuri** [comment]
I have tried to make changes. Let me know if there are improvements

**Comment by 0x5bfa** [comment]
Thanks for your contribution. You can rewrite ‘Closes #issues…’ to ‘Closes #11697’.

And are you sure there’s nothing suspicious behavior? Interaction test appears failed, which means some items are not accessible or a crash happened.

**Comment by gave92** [comment]
The copypaste test would need to be updated because it checks for a file named "- Copy".

**Comment by 0x5bfa** [comment]
Please modify here too

https://github.com/files-community/Files/blob/516ceb27b0dd7d6aeb3876c1f84d2d3f81249457/tests/Files.InteractionTests/Tests/FolderTests.cs#L142

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 1]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
 				}
 			}
 		}
+
+		public static string GetIncrementalName(bool overWriteOnCopy,string? filePathToCheck)
+		{
+			if (filePathToCheck == null)
+				return null;
+			else
+			{
```
Comment (QuaintMako) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1027]: `Else` keyword can be removed to gain a level of depth.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 2]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
...
+				if ((!Path.Exists(filePathToCheck)) || overWriteOnCopy)
+					return Path.GetFileName(filePathToCheck);
+				else
```
Comment (QuaintMako) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1030]: `Else` keyword can be removed, the `return` in the `if` is enough.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 3]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
 				}
 			}
 		}
+
+		public static string GetIncrementalName(bool overWriteOnCopy,string? filePathToCheck)
```
Comment (QuaintMako) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1022]: A space is needed after the coma:
```suggestion
		public static string GetIncrementalName(bool overWriteOnCopy, string? filePathToCheck)
```

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 4]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
...
+			if (filePathToCheck == null)
+				return null;
+			else
+			{
+				if ((!Path.Exists(filePathToCheck)) || overWriteOnCopy)
+					return Path.GetFileName(filePathToCheck);
+				else
+				{
+					int i = 2;
+					string genfilepath = string.Concat([filePathToCheck, " - (", i.ToString(), ")"]);
```
Comment (QuaintMako) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1033]: `string.Concat([filePathToCheck, " - (", i.ToString(), ")"]);` could be wrapped into a lambda to be called twice to reduce duplication.

Comment (QuaintMako) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1033]: `genfilepath` should be `genFilePath` to stay consistent with naming.

**Review by QuaintMako — COMMENTED** [review]
Few feedback on the formatting! :)

**Comment by hishitetsu** [comment]
* Your PR appends the number after the file extension, but it should be appended before the file extension.
* If you want to match the behavior with Explorer, the hyphen in front of the number is not necessary.
* This change should only be applied when copying files to a folder different from the source. The behavior when pasting into the same folder should not be changed from the current behavior. (- Copy, - Copy (2), - Copy (3), ...)

**Comment by ashrafmansuri** [comment]
> Few feedback on the formatting! :)

Thank you :) working on it as soon as I get home.

**Comment by ashrafmansuri** [comment]
> * Your PR appends the number after the file extension, but it should be appended before the file extension.
> 
> * If you want to match the behavior with Explorer, the hyphen in front of the number is not necessary.
> 
> * This change should only be applied when copying files to a folder different from the source. The behavior when pasting into the same folder should not be changed from the current behavior. (- Copy, - Copy (2), - Copy (3), ...)

Understood! Gonna make the necessary changes :)

**Comment by ashrafmansuri** [comment]
I have made a commit. Looks fine for a file but not a directory. working on it

**Comment by ashrafmansuri** [comment]
The last commit seems to be working as expected. Please review and let me know if there are any suggestions. Thank you :)

**Review by ashrafmansuri — COMMENTED** [review]
Thank you for the suggestions :)

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 5]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
...
+
+		public static string GetIncrementalName(bool overWriteOnCopy, string? filePathToCheck, string? filePathToCopy)
+		{
+			if (filePathToCheck == null)
+				return null;			
+
+			if ((!Path.Exists(filePathToCheck)) || overWriteOnCopy || filePathToCheck == filePathToCopy)
+				return Path.GetFileName(filePathToCheck);
+
+			int i = 2;
```
Comment (0x5bfa) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1030]: i => index

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 6]
```diff
@@ -1018,5 +1018,27 @@ protected override void Dispose(bool disposing)
...
+			string filePath = filePathToCheck;
+			if (Path.HasExtension(filePathToCheck))
+				filePath = filePathToCheck.Substring(0, filePathToCheck.LastIndexOf("."));
+
+			Func<int, string> genFilePath = x => string.Concat([filePath, " (", x.ToString(), ")", Path.GetExtension(filePathToCheck)]);
+
+			while (Path.HasExtension(filePathToCheck) ? File.Exists(genFilePath(i)) : Path.Exists(genFilePath(i)))
+			{
```
Comment (0x5bfa) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1038]: 1. Remove brackets and insert a line break right before `return` statement.
2. Let's use `index++;` instead.

**Review by 0x5bfa — COMMENTED** [review]
Some suggestions for codebase quality.

Here's too:
https://github.com/files-community/Files/pull/15267#issuecomment-2081480219

**Comment by ashrafmansuri** [comment]
Test might be failing because it apparently copies in same directory. So by default it appends - Copy in the end.

**Review by 0x5bfa — APPROVED** [review]
LGTM code wise, thanks.

**Comment by hishitetsu** [comment]
I'll review and test by the end of this week.

**Comment by yaira2** [comment]
Overall this is working as expected but there is one issue that I think should be addressed before we merge
1. Select generate new name in the conflicts dialog
2. Focus the edit name TextBox
3. Click continue
4. Notice that the generated number is incorrect
image

**Comment by ashrafmansuri** [comment]
> Overall this is working as expected but there is one issue that I think should be addressed before we merge
> 
> 1. Select generate new name in the conflicts dialog
> 
> 2. Focus the edit name TextBox
> 
> 3. Click continue
> 
> 4. Notice that the generated number is incorrect
> 
> image
> 
> 

Okay.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 7]
```diff
@@ -1018,5 +1018,26 @@ protected override void Dispose(bool disposing)
...
+				return Path.GetFileName(filePathToCheck);
+
+			int index = 2;
+			string filePath = filePathToCheck;
+			if (Path.HasExtension(filePathToCheck))
+				filePath = filePathToCheck.Substring(0, filePathToCheck.LastIndexOf("."));
+
+			Func<int, string> genFilePath = x => string.Concat([filePath, " (", x.ToString(), ")", Path.GetExtension(filePathToCheck)]);
+
+			while (Path.HasExtension(filePathToCheck) ? File.Exists(genFilePath(index)) : Path.Exists(genFilePath(index)))
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1037]:
```suggestion
			while (Path.Exists(genFilePath(index)))			
```
Some files don't have extensions, and it is enough to just use `Path.Exists()` in the first place.

**Comment by ashrafmansuri** [comment]
> Overall this is working as expected but there is one issue that I think should be addressed before we merge
> 
> 1. Select generate new name in the conflicts dialog
> 2. Focus the edit name TextBox
> 3. Click continue
> 4. Notice that the generated number is incorrect
>    image

The file copies with the new name but gets renamed.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 8]
```diff
@@ -580,9 +580,10 @@ public static Task<(bool, ShellOperationResult)> CopyItemAsync(string[] fileToCo
 					{
 						using ShellItem shi = new(fileToCopyPath[i]);
 						using ShellFolder shd = new(Path.GetDirectoryName(copyDestination[i]));
-
+						string fileName = GetIncrementalName(overwriteOnCopy, copyDestination[i], fileToCopyPath[i]);
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:583]:
```suggestion
						var fileName = GetIncrementalName(overwriteOnCopy, copyDestination[i], fileToCopyPath[i]);

```
Use `var` whenever possible.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 9]
```diff
@@ -580,9 +580,10 @@ public static Task<(bool, ShellOperationResult)> CopyItemAsync(string[] fileToCo
...
 						// Performa copy operation
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:584]:
```suggestion
						// Perform copy operation
```
A typo. It's there originally, though.

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 10]
```diff
@@ -580,9 +580,10 @@ public static Task<(bool, ShellOperationResult)> CopyItemAsync(string[] fileToCo
...
-						op.QueueCopyOperation(shi, shd, Path.GetFileName(copyDestination[i]));
+						op.QueueCopyOperation(shi, shd, fileName);
+						//
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:586]:
```suggestion
						op.QueueCopyOperation(shi, shd, fileName);
```

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 11]
```diff
@@ -1018,5 +1019,26 @@ protected override void Dispose(bool disposing)
...
+		public static string GetIncrementalName(bool overWriteOnCopy, string? filePathToCheck, string? filePathToCopy)
+		{
+			if (filePathToCheck == null)
+				return null;			
+
+			if ((!Path.Exists(filePathToCheck)) || overWriteOnCopy || filePathToCheck == filePathToCopy)
+				return Path.GetFileName(filePathToCheck);
+
+			int index = 2;
+			string filePath = filePathToCheck;
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1032]:
```suggestion
			var index = 2;
			var filePath = filePathToCheck;
```

**Review by hishitetsu — COMMENTED** [review]
It works well! There are a few more things in terms of code quality.

**Comment by ashrafmansuri** [comment]
> It works well! There are a few more things in terms of code quality.

I was thinking whether I should declare a variable or directly call the function? My only deciding factor was declaring variable in this case makes it more easy to read. What do you suggest?

**Comment by hishitetsu** [comment]
> I was thinking whether I should declare a variable or directly call the function? My only deciding factor was declaring variable in this case makes it more easy to read. What do you suggest?

Are you talking about `Func<int, string> genFilePath`? I think it's fine.

**Comment by ashrafmansuri** [comment]
> > I was thinking whether I should declare a variable or directly call the function? My only deciding factor was declaring variable in this case makes it more easy to read. What do you suggest?
> 
> 
> 
> Are you talking about `Func<int, string> genFilePath`? I think it's fine.

Oh no no I was saying about var fileName.

**Comment by hishitetsu** [comment]
> Oh no no I was saying about var fileName.

There is no problem as it is 👍

**Comment by ashrafmansuri** [comment]
> > Oh no no I was saying about var fileName.
> 
> There is no problem as it is 👍

Alright! Thank you so much! I have committed suggested changes!

File: src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs [hunk 12]
```diff
@@ -1018,5 +1020,26 @@ protected override void Dispose(bool disposing)
 				}
 			}
 		}
+
+		public static string GetIncrementalName(bool overWriteOnCopy, string? filePathToCheck, string? filePathToCopy)
```
Comment (hishitetsu) [inline src/Files.App/Utils/Storage/Operations/FileOperationsHelpers.cs:1024]:
```suggestion
		private static string GetIncrementalName(bool overWriteOnCopy, string? filePathToCheck, string? filePathToCopy)
```
Sorry, I forgot one comment.
It should be `private` if it doesn't need to be public.

**Review by hishitetsu — APPROVED** [review]
Thank you! LGTM

**Comment by hishitetsu** [comment]
@ashrafmansuri there are lines with extra spaces, please remove them if you have time.
image

**Comment by ashrafmansuri** [comment]
> @ashrafmansuri there are lines with extra spaces, please remove them if you have time. image

Done! Thanks! 👍

**Review by yaira2 — APPROVED** [review]
Nice work! Thank you for contributing to Files.

**Comment by ashrafmansuri** [comment]
> Nice work! Thank you for contributing to Files.

Thank you! My pleasure! ✌🏻

**Pull request merged**